### PR TITLE
Have AddImport compute a pure-data "fix" object that will be fed into the right code action.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/19893

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Pilchie 

This is the next stage of Add-Using oop that follows up #19893.  In this PR we introduce a data type that stores the information computed by the analysis-portion of add-using and passes it along to the application-portion of add-using.

With this change we can then move the analysis-portion of add-using OOP and pass this data result back to VS for the application-portion to use.

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Tags;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.AddImport
+{
+    internal enum AddImportFixKind
```
Comment (sharwell) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: :bulb: Move to **AddImportFixKind.cs**

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        /// Tags that control what glyph is displayed in the lightbulb menu.
+        /// </summary>
+        public ImmutableArray<string> Tags { get; private set; }
+
+        /// <summary>
+        /// The priority this item should have in the lightbulb list.
+        /// </summary>
+        public CodeActionPriority Priority { get; private set; }
+
+        #region When adding P2P refrences.
```
Comment (sharwell) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: :memo: Very much not a fan of having regions for this. Either of the following would be preferable:

1. Using derived types, assuming the serialization layer supports it.
2. Using a `<remarks>` element on each property, like this for the current one.

    ```csharp
    /// <remarks>
    /// <para>This property is only used when <see cref="Kind"/> is
    /// <see cref="AddImportFixKind.ProjectSymbol"/>.</para>
    /// </remarks>
    ```

File: src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs [hunk 3]
```diff
@@ -540,7 +540,7 @@ public IEnumerable<Project> GetProjectsWithInstalledPackage(Solution solution, s
             {
                 var state = kvp.Value;
                 var versionSet = state.InstalledPackageToVersion[packageName];
-                if (versionSet.Contains(packageName))
+                if (versionSet.Contains(version))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs]: :memo: This change seems unrelated to the rest of the work.

File: src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs [hunk 4]
```diff
@@ -17,22 +14,13 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
                 private readonly Lazy<string> _lazyResolvedPath;
 
                 public AssemblyReferenceCodeAction(
                     Document originalDocument,
-                    ImmutableArray<TextChange> textChanges,
-                    string title,
-                    string assemblyName,
-                    string fullyQualifiedTypeName)
-                    : base(originalDocument, textChanges, title, WellKnownTagArrays.AddReference, CodeActionPriority.Low)
+                    AddImportFixData fixData)
```
Comment (sharwell) [inline src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs]: :bulb: If you don't use types derived from `AddImportFixData`, each of these code actions could benefit from a 'requires' like for the `fixData.Kind`.

File: src/Features/Core/Portable/AddImport/CodeActions/AddImportCodeAction.cs [hunk 5]
```diff
@@ -36,15 +38,16 @@ private abstract class AddImportCodeAction : CodeAction
 
             protected AddImportCodeAction(
                 Document originalDocument,
-                ImmutableArray<TextChange> textChanges,
-                string title, ImmutableArray<string> tags,
-                CodeActionPriority priority)
+                AddImportFixData fixData,
+                string titleOpt = null)
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddImport/CodeActions/AddImportCodeAction.cs]: Still not a fan of optional parameters on internal APIs.

**Comment by CyrusNajmabadi** [comment]
> LGTM, though I was wondering - don't we also need to move the FullyQualify fix stuff over to OOP in order to avoid building indices?

Fortunately no.  Fully-qualify doesn't search outside of the current project.  It only ever searches the current project.  And it does so in a way that is already remoted.  The remote call will return a symbol back to the VS side, btu that's cheap in the case when you're calling on the current project (since we already have the compilations and whatnot ready).

The expensive cases would be if the OOP side returned results for *other* projects that we'd then need to create compilations for.

**Comment by CyrusNajmabadi** [comment]
https://ci.dot.net/job/dotnet_roslyn/job/master/job/perf_correctness_prtest/6950/ and https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/4955/ are both nuget cache corruption.  retesting.

**Comment by CyrusNajmabadi** [comment]
retest perf_correctness_prtest  please

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest  please

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: Sure.

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: Using derived types is currently a PITA in the serialization layer.  Have been talking to heejae about improving that.  If/when we improve things there we can move over.  

> Using a ```<remarks>``` element on each property, like this for the current one.

I'm actually very opposed to that for the reason that comments invariably become desynced from code organization.   regions are part of the code and it is immediately visible when a new field is placed into the wrong region.

File: src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs]: Sure.  will add.

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/CodeActions/AssemblyReferenceCodeAction.cs]: #fixed

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: #fixed

File: src/Features/Core/Portable/AddImport/AddImportFixData.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AddImportFixData.cs]: #KeepingItReal

File: src/Features/Core/Portable/AddImport/CodeActions/AddImportCodeAction.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/CodeActions/AddImportCodeAction.cs]: Will change.

**Comment by CyrusNajmabadi** [comment]
Telling @MattGertz  about this.  This is the next step in OOPifying add-using.  In this PR we introduce a data type that stores the information computed by the analysis-portion of add-using and passes it along to the application-portion of add-using.

With this change we can then move the analysis-portion of add-using OOP and pass this data result back to VS for the application-portion to use.

**Comment by CyrusNajmabadi** [comment]
integration tests aborted: https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/4974/

**Comment by CyrusNajmabadi** [comment]
retest windows_release_vs-integration_prtest  please

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
retest windows_release_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
A followup PR to this was merged in already.
